jwst/assign_wcs/nircam.py

```diff
@@ -54,14 +70,271 @@ def imaging_distortion(input_model, reference_files):
+    catalog in units of RA,DEC so they can be translated to pixels by
+    the dispersed image's imaging wcs.
+
+    For each spectral order, the configuration file contains a
+    magnitude-cutoff value. Sources with magnitudes fainter than the
+    extraction cutoff (MMAG_EXTRACT)  will not be extracted, but are
+    accounted for when computing the spectral contamination and background
+    estimates. The default extraction value is 99 right now.
+
+    The sensitivity information from the original aXe style configuration
+    file needs to be modified by the passband of the filter used for
+    the direct image to get the min and max wavelengths
+    which correspond to t=0 and t=1, this currently has been done by the team
+    and the min and max wavelengths to use to calculate t are stored in the
+    grism reference file as wrange, which can be selected by wrange_selector
+    which contains the filter names.
+
+
+    Step 1: Convert the source catalog from the reference frame of the
+            uberimage to that of the dispersed image.  For the Vanilla
+            Pipeline we assume that the pointing information in the file
+            headers is sufficient.  This will be strictly true if all images
+            were obtained in a single visit (same guide stars).
+    Step 2: Record source information for each object in the catalog: position
+            (RA and Dec), shape (A_IMAGE, B_IMAGE, THETA_IMAGE), and all
+            available magnitudes.
+    Step 3: Compute the trace and wavelength solutions for each object in the
+            catalog and for each spectral order.  Record this information.
+    Step 4: Compute the WIDTH of each spectral subwindow, which may be fixed or
+            variable (see discussion of optimal extraction, below).  Record
+            this information.
+    Step 4: Record the MMAG_EXTRACT for each object and spectral order.
+
+    """
+
+    # The input is the grism image
+    if not isinstance(input_model, ImageModel):
+        raise TypeError('The input data model must be an ImageModel.')
+
+    # make sure this is a grism image
+    if "NRC_GRISM" not in input_model.meta.exposure.type:
+            raise TypeError('The input exposure is not a NIRCAM grism')
+
+    # Create the empty detector as a 2D coordinate frame in pixel units
+    gdetector = cf.Frame2D(name='grism_detector',
+                           axes_order=(0, 1),
+                           unit=(u.pix, u.pix))
+
+    # translate the x,y detector-in to x,y detector out coordinates
+    # Get the disperser parameters which are defined as a model for each
+    # spectral order
+    log.info("Collecting specwcs information")
+    with NIRCAMGrismModel(reference_files['specwcs']) as f:
+        xcoeff = f.xcoeff
+        ycoeff = f.ycoeff
+        lcoeff = f.lcoeff
+        orders = [j for j, k in enumerate(f.orders)]  # ordered [str names]
+        wrange_selector = f.wrange_selector
```

Comment:
Does it make sense to have the wavelengthrange information in a separate ref file? We already use `wavelengthrange` reftype for MIRI and Nirspec.